File: src/ui_python/libsnes.py
import sys, time, array, ctypes, os

libsnes = ctypes.cdll.LoadLibrary("/usr/local/lib/libsnes.so")

snes_video_refresh_t = ctypes.CFUNCTYPE(None, ctypes.POINTER(ctypes.c_uint16), ctypes.c_uint, ctypes.c_uint)
snes_set_video_refresh = libsnes.snes_set_video_refresh
snes_set_video_refresh.restype = None
snes_set_video_refresh.argtypes = [snes_video_refresh_t]

snes_audio_sample_t = ctypes.CFUNCTYPE(None, ctypes.c_uint16, ctypes.c_uint16)
snes_set_audio_sample = libsnes.snes_set_audio_sample
snes_set_audio_sample.restype = None
snes_set_audio_sample.argtypes = [snes_audio_sample_t]

snes_input_poll_t = ctypes.CFUNCTYPE(None)
snes_set_input_poll = libsnes.snes_set_input_poll
snes_set_input_poll.restype = None
snes_set_input_poll.argtypes = [snes_input_poll_t]

snes_input_state_t = ctypes.CFUNCTYPE(ctypes.c_int16, ctypes.c_bool, ctypes.c_uint, ctypes.c_uint, ctypes.c_uint)
snes_set_input_state = libsnes.snes_set_input_state
snes_set_input_state.restype = ctypes.c_int16
snes_set_input_state.argtypes = [snes_input_state_t]

snes_library_revision = libsnes.snes_library_revision
snes_library_revision.restype = ctypes.c_uint
snes_library_revision.argtypes = []

snes_init = libsnes.snes_init
snes_init.restype = None
snes_init.argtypes = []

snes_term = libsnes.snes_term
snes_term.restype = None
snes_term.argtypes = []

snes_power = libsnes.snes_power
snes_power.restype = None
snes_power.argtypes = []

snes_reset = libsnes.snes_reset
snes_reset.restype = None
snes_reset.argtypes = []

snes_run = libsnes.snes_run
snes_run.restype = None
snes_run.argtypes = []

snes_unload = libsnes.snes_unload
snes_unload.restype = None
snes_unload.argtypes = []

snes_serialize_size = libsnes.snes_serialize_size
snes_serialize_size.restype = ctypes.c_uint
snes_serialize_size.argtypes = []

snes_serialize = libsnes.snes_serialize
snes_serialize.restype = ctypes.c_bool
snes_serialize.argtypes = [ctypes.POINTER(ctypes.c_uint8), ctypes.c_uint]

snes_unserialize = libsnes.snes_unserialize
snes_unserialize.restype = ctypes.c_bool
snes_unserialize.argtypes = [ctypes.POINTER(ctypes.c_uint8), ctypes.c_uint]

snes_cheat_reset = libsnes.snes_cheat_reset
snes_cheat_reset.restype = None
snes_cheat_reset.argtypes = []

snes_cheat_set = libsnes.snes_cheat_set
snes_cheat_set.restype = None
snes_cheat_set.argtypes = [ctypes.c_uint, ctypes.c_bool, ctypes.POINTER(ctypes.c_char)]

snes_load_cartridge_normal = libsnes.snes_load_cartridge_normal
snes_load_cartridge_normal.restype = None
snes_load_cartridge_normal.argtypes = [ctypes.POINTER(ctypes.c_char), ctypes.POINTER(ctypes.c_uint8), ctypes.c_uint]

snes_get_region = libsnes.snes_get_region
snes_get_region.restype = ctypes.c_bool
snes_get_region.argtypes = []

snes_get_memory_data = libsnes.snes_get_memory_data
snes_get_memory_data.restype = ctypes.POINTER(ctypes.c_uint8)
snes_get_memory_data.argtypes = [ctypes.c_uint]

snes_get_memory_size = libsnes.snes_get_memory_size
snes_get_memory_size.restype = ctypes.c_uint
snes_get_memory_size.argtypes = [ctypes.c_uint]

snes_blit_set_colortable = libsnes.snes_blit_set_colortable
snes_blit_set_colortable.restype = None
snes_blit_set_colortable.argtypes = [ctypes.POINTER(ctypes.c_uint32)]

snes_blit = libsnes.snes_blit
snes_blit.restype = None
snes_blit.argtypes = [ctypes.POINTER(ctypes.c_uint32), ctypes.c_uint, ctypes.c_uint, ctypes.c_uint, ctypes.POINTER(ctypes.c_uint16), ctypes.c_uint, ctypes.c_uint, ctypes.c_uint]

class SNES:
	def __init__(self):
		self.loaded = False

	def library_revision(self):
		return snes_library_revision()

	def set_video_refresh(self, callback):
		self.video_refresh = snes_video_refresh_t(callback)
		snes_set_video_refresh(self.video_refresh)

	def set_audio_sample(self, callback):
		self.audio_sample = snes_audio_sample_t(callback)
		snes_set_audio_sample(self.audio_sample)

	def set_input_poll(self, callback):
		self.input_poll = snes_input_poll_t(callback)
		snes_set_input_poll(self.input_poll)

	def set_input_state(self, callback):
		self.input_state = snes_input_state_t(callback)
		snes_set_input_state(self.input_state)

	def init(self):
		snes_init()

	def term(self):
		snes_term()

	def power(self):
		snes_power()

	def reset(self):
		snes_reset()

	def run(self):
		snes_run()

	def unload(self):
		snes_unload()
		self.loaded = False

	def serialize_size(self):
		return snes_serialize_size()

	def load_cartridge_normal(self, filename):
		size = os.path.getsize(filename)
		fp = open(filename, "rb")
		data = fp.read()
		snes_load_cartridge_normal(None, ctypes.cast(data, ctypes.POINTER(ctypes.c_uint8)), size)
		fp.close()
		self.loaded = True

	def get_region(self):
		return snes_get_region()

	def set_colortable(self, colortable):
		snes_blit_set_colortable(ctypes.cast(colortable.buffer_info()[0], ctypes.POINTER(ctypes.c_uint32)))

	def blit(self, outbuffer, outpitch, outwidth, outheight, inbuffer, inpitch, width, height):
		snes_blit(ctypes.cast(outbuffer.buffer_info()[0], ctypes.POINTER(ctypes.c_uint32)), outpitch, outwidth, outheight, inbuffer, inpitch, width, height)

File: bsnes/ui-libsnes/libsnes.cpp
#include "libsnes.hpp"
#include <snes/snes.hpp>

#include <nall/snes/cartridge.hpp>
#include <nall/gameboy/cartridge.hpp>
using namespace nall;

struct Interface : public SNES::Interface {
  snes_video_refresh_t pvideo_refresh;
  snes_audio_sample_t paudio_sample;
  snes_input_poll_t pinput_poll;
  snes_input_state_t pinput_state;
  string basename;

  void video_refresh(const uint16_t *data, unsigned width, unsigned height) {
    if(pvideo_refresh) return pvideo_refresh(data, width, height);
  }

  void audio_sample(uint16_t left, uint16_t right) {
    if(paudio_sample) return paudio_sample(left, right);
  }

  void input_poll() {
    if(pinput_poll) return pinput_poll();
  }

  int16_t input_poll(bool port, SNES::Input::Device device, unsigned index, unsigned id) {
    if(pinput_state) return pinput_state(port, (unsigned)device, index, id);
    return 0;
  }

  void message(const string &text) {
    print(text, "\n");
  }

  string path(SNES::Cartridge::Slot slot, const string &hint) {
    return { basename, hint };
  }

  Interface() : pvideo_refresh(0), paudio_sample(0), pinput_poll(0), pinput_state(0) {
  }
};

static Interface interface;

const char* snes_library_id(void) {
  static string id = { SNES::Info::Name, " v", SNES::Info::Version };
  return (const char*)id;
}

unsigned snes_library_revision_major(void) {
  return 1;
}

unsigned snes_library_revision_minor(void) {
  return 3;
}

void snes_set_video_refresh(snes_video_refresh_t video_refresh) {
  interface.pvideo_refresh = video_refresh;
}

void snes_set_audio_sample(snes_audio_sample_t audio_sample) {
  interface.paudio_sample = audio_sample;
}

void snes_set_input_poll(snes_input_poll_t input_poll) {
  interface.pinput_poll = input_poll;
}

void snes_set_input_state(snes_input_state_t input_state) {
  interface.pinput_state = input_state;
}

void snes_set_controller_port_device(bool port, unsigned device) {
  SNES::input.port_set_device(port, (SNES::Input::Device)device);
}

void snes_set_cartridge_basename(const char *basename) {
  interface.basename = basename;
}

void snes_init(void) {
  SNES::system.init(&interface);
  SNES::input.port_set_device(0, SNES::Input::Device::Joypad);
  SNES::input.port_set_device(1, SNES::Input::Device::Joypad);
}

void snes_term(void) {
  SNES::system.term();
}

void snes_power(void) {
  SNES::system.power();
}

void snes_reset(void) {
  SNES::system.reset();
}

void snes_run(void) {
  SNES::system.run();
}

unsigned snes_serialize_size(void) {
  return SNES::system.serialize_size();
}

bool snes_serialize(uint8_t *data, unsigned size) {
  SNES::system.runtosave();
  serializer s = SNES::system.serialize();
  if(s.size() > size) return false;
  memcpy(data, s.data(), s.size());
  return true;
}

bool snes_unserialize(const uint8_t *data, unsigned size) {
  serializer s(data, size);
  return SNES::system.unserialize(s);
}

void snes_cheat_reset(void) {
  SNES::cheat.reset();
  SNES::cheat.synchronize();
}

void snes_cheat_set(unsigned index, bool enabled, const char *code) {
  SNES::cheat[index] = code;
  SNES::cheat[index].enabled = enabled;
  SNES::cheat.synchronize();
}

bool snes_load_cartridge_normal(
  const char *rom_xml, const uint8_t *rom_data, unsigned rom_size
) {
  snes_cheat_reset();
  if(rom_data) SNES::cartridge.rom.copy(rom_data, rom_size);
  string xmlrom = (rom_xml && *rom_xml) ? string(rom_xml) : SNESCartridge(rom_data, rom_size).xmlMemoryMap;
  SNES::cartridge.load(SNES::Cartridge::Mode::Normal, { xmlrom });
  SNES::system.power();
  return true;
}

bool snes_load_cartridge_bsx_slotted(
  const char *rom_xml, const uint8_t *rom_data, unsigned rom_size,
  const char *bsx_xml, const uint8_t *bsx_data, unsigned bsx_size
) {
  snes_cheat_reset();
  if(rom_data) SNES::cartridge.rom.copy(rom_data, rom_size);
  string xmlrom = (rom_xml && *rom_xml) ? string(rom_xml) : SNESCartridge(rom_data, rom_size).xmlMemoryMap;
  if(bsx_data) SNES::bsxflash.memory.copy(bsx_data, bsx_size);
  string xmlbsx = (bsx_xml && *bsx_xml) ? string(bsx_xml) : SNESCartridge(bsx_data, bsx_size).xmlMemoryMap;
  SNES::cartridge.load(SNES::Cartridge::Mode::BsxSlotted, { xmlrom, xmlbsx });
  SNES::system.power();
  return true;
}

bool snes_load_cartridge_bsx(
  const char *rom_xml, const uint8_t *rom_data, unsigned rom_size,
  const char *bsx_xml, const uint8_t *bsx_data, unsigned bsx_size
) {
  snes_cheat_reset();
  if(rom_data) SNES::cartridge.rom.copy(rom_data, rom_size);
  string xmlrom = (rom_xml && *rom_xml) ? string(rom_xml) : SNESCartridge(rom_data, rom_size).xmlMemoryMap;
  if(bsx_data) SNES::bsxflash.memory.copy(bsx_data, bsx_size);
  string xmlbsx = (bsx_xml && *bsx_xml) ? string(bsx_xml) : SNESCartridge(bsx_data, bsx_size).xmlMemoryMap;
  SNES::cartridge.load(SNES::Cartridge::Mode::Bsx, { xmlrom, xmlbsx });
  SNES::system.power();
  return true;
}

bool snes_load_cartridge_sufami_turbo(
  const char *rom_xml, const uint8_t *rom_data, unsigned rom_size,
  const char *sta_xml, const uint8_t *sta_data, unsigned sta_size,
  const char *stb_xml, const uint8_t *stb_data, unsigned stb_size
) {
  snes_cheat_reset();
  if(rom_data) SNES::cartridge.rom.copy(rom_data, rom_size);
  string xmlrom = (rom_xml && *rom_xml) ? string(rom_xml) : SNESCartridge(rom_data, rom_size).xmlMemoryMap;
  if(sta_data) SNES::sufamiturbo.slotA.rom.copy(sta_data, sta_size);
  string xmlsta = (sta_xml && *sta_xml) ? string(sta_xml) : SNESCartridge(sta_data, sta_size).xmlMemoryMap;
  if(stb_data) SNES::sufamiturbo.slotB.rom.copy(stb_data, stb_size);
  string xmlstb = (stb_xml && *stb_xml) ? string(stb_xml) : SNESCartridge(stb_data, stb_size).xmlMemoryMap;
  SNES::cartridge.load(SNES::Cartridge::Mode::SufamiTurbo, { xmlrom, xmlsta, xmlstb });
  SNES::system.power();
  return true;
}

bool snes_load_cartridge_super_game_boy(
  const char *rom_xml, const uint8_t *rom_data, unsigned rom_size,
  const char *dmg_xml, const uint8_t *dmg_data, unsigned dmg_size
) {
  snes_cheat_reset();
  if(rom_data) SNES::cartridge.rom.copy(rom_data, rom_size);
  string xmlrom = (rom_xml && *rom_xml) ? string(rom_xml) : SNESCartridge(rom_data, rom_size).xmlMemoryMap;
  if(dmg_data) {
    string xmldmg = (dmg_xml && *dmg_xml) ? string(dmg_xml) : GameBoyCartridge(dmg_data, dmg_size).xml;
    GameBoy::cartridge.load(xmldmg, dmg_data, dmg_size);
  }
  SNES::cartridge.load(SNES::Cartridge::Mode::SuperGameBoy, { xmlrom, "" });
  SNES::system.power();
  return true;
}

void snes_unload_cartridge(void) {
  SNES::cartridge.unload();
}

bool snes_get_region(void) {
  return SNES::system.region() == SNES::System::Region::NTSC ? 0 : 1;
}

uint8_t* snes_get_memory_data(unsigned id) {
  if(SNES::cartridge.loaded() == false) return 0;

  switch(id) {
    case SNES_MEMORY_CARTRIDGE_RAM:
      return SNES::cartridge.ram.data();
    case SNES_MEMORY_CARTRIDGE_RTC:
      if(SNES::cartridge.has_srtc()) return SNES::srtc.rtc;
      if(SNES::cartridge.has_spc7110rtc()) return SNES::spc7110.rtc;
      return 0;
    case SNES_MEMORY_BSX_RAM:
      if(SNES::cartridge.mode() != SNES::Cartridge::Mode::Bsx) break;
      return SNES::bsxcartridge.sram.data();
    case SNES_MEMORY_BSX_PRAM:
      if(SNES::cartridge.mode() != SNES::Cartridge::Mode::Bsx) break;
      return SNES::bsxcartridge.psram.data();
    case SNES_MEMORY_SUFAMI_TURBO_A_RAM:
      if(SNES::cartridge.mode() != SNES::Cartridge::Mode::SufamiTurbo) break;
      return SNES::sufamiturbo.slotA.ram.data();
    case SNES_MEMORY_SUFAMI_TURBO_B_RAM:
      if(SNES::cartridge.mode() != SNES::Cartridge::Mode::SufamiTurbo) break;
      return SNES::sufamiturbo.slotB.ram.data();
    case SNES_MEMORY_GAME_BOY_RAM:
      if(SNES::cartridge.mode() != SNES::Cartridge::Mode::SuperGameBoy) break;
      return GameBoy::cartridge.ramdata;
  //case SNES_MEMORY_GAME_BOY_RTC:
  //  if(SNES::cartridge.mode() != SNES::Cartridge::Mode::SuperGameBoy) break;
  //  return GameBoy::cartridge.rtcdata;

    case SNES_MEMORY_WRAM:
      return SNES::cpu.wram;
    case SNES_MEMORY_APURAM:
      return SNES::smp.apuram;
    case SNES_MEMORY_VRAM:
      return SNES::ppu.vram;
    case SNES_MEMORY_OAM:
      return SNES::ppu.oam;
    case SNES_MEMORY_CGRAM:
      return SNES::ppu.cgram;
  }

  return 0;
}

unsigned snes_get_memory_size(unsigned id) {
  if(SNES::cartridge.loaded() == false) return 0;
  unsigned size = 0;

  switch(id) {
    case SNES_MEMORY_CARTRIDGE_RAM:
      size = SNES::cartridge.ram.size();
      break;
    case SNES_MEMORY_CARTRIDGE_RTC:
      if(SNES::cartridge.has_srtc() || SNES::cartridge.has_spc7110rtc()) size = 20;
      break;
    case SNES_MEMORY_BSX_RAM:
      if(SNES::cartridge.mode() != SNES::Cartridge::Mode::Bsx) break;
      size = SNES::bsxcartridge.sram.size();
      break;
    case SNES_MEMORY_BSX_PRAM:
      if(SNES::cartridge.mode() != SNES::Cartridge::Mode::Bsx) break;
      size = SNES::bsxcartridge.psram.size();
      break;
    case SNES_MEMORY_SUFAMI_TURBO_A_RAM:
      if(SNES::cartridge.mode() != SNES::Cartridge::Mode::SufamiTurbo) break;
      size = SNES::sufamiturbo.slotA.ram.size();
      break;
    case SNES_MEMORY_SUFAMI_TURBO_B_RAM:
      if(SNES::cartridge.mode() != SNES::Cartridge::Mode::SufamiTurbo) break;
      size = SNES::sufamiturbo.slotB.ram.size();
      break;
    case SNES_MEMORY_GAME_BOY_RAM:
      if(SNES::cartridge.mode() != SNES::Cartridge::Mode::SuperGameBoy) break;
      size = GameBoy::cartridge.ramsize;
      break;
  //case SNES_MEMORY_GAME_BOY_RTC:
  //  if(SNES::cartridge.mode() != SNES::Cartridge::Mode::SuperGameBoy) break;
  //  size = GameBoy::cartridge.rtcsize;
  //  break;

    case SNES_MEMORY_WRAM:
      size = 128 * 1024;
      break;
    case SNES_MEMORY_APURAM:
      size = 64 * 1024;
      break;
    case SNES_MEMORY_VRAM:
      size = 64 * 1024;
      break;
    case SNES_MEMORY_OAM:
      size = 544;
      break;
    case SNES_MEMORY_CGRAM:
      size = 512;
      break;
  }

  if(size == -1U) size = 0;
  return size;
}
